Domain Picker: Added "Use a domain I own" item

tjcafferkey opened the pull request:
#### Changes proposed in this Pull Request

* Adds a new component for "Use a domain I own" in the `<DomainPicker />` component.
* This links the user to `/start/new-launch/domains-launch/use-your-domain`

#### Testing instructions

Setup:
1. Checkout this branch
2. Run `yarn start` in Claypso
2. Run `yarn dev --sync` in `apps/editing-toolkit`
3. Run `yarn dev --sync` in `apps/wpcom-block-editor`
4. Sandbox an unlaunched site created via `/start`

**Focused launch**

1. Go to `/page/UNLAUNCHED_SITE_ID_CREATED_WITH_START.wordpress.com/home?flags=create/focused-launch-flow`
2. Click "Launch" and confirm that "Use a domain I own" is not present in the Summary view
3. Click "View all domains" and confirm "Use a domain I own" is present and links to `/start/new-launch/domains-launch/use-your-domain?siteSlug=YOUR_SITE_SLUG&source=YOUR_FOCUSED_LAUNCH_URL`

`YOUR_FOCUSED_LAUNCH_URL` should be the absolute URL from Calypso where you were viewing focused launch.

- [ ] Domain I own component is present
- [ ] Links to `/start/new-launch/domains-launch/use-your-domain?siteSlug=YOUR_SITE_SLUG&source=YOUR_FOCUSED_LAUNCH_URL`
- [ ] Clicking "Map your domain", and then "Back" takes you back to the original step
- [ ] Clicking "Back" on the original step takes you back to the editor you came from (ie. the `source` param value)

**Gutenboarding - Regression testing**

1. Go to `/new`.
2. Enter a site name. Go to next step which should be the domain step.
3. Confirm "Use a domain I own" is not present in Gutenboarding domain picker.

- [ ] "Use a domain I own" component is not present

**Editing Toolkit Step-by-Step Launch - Regression testing**

1. Checkout this branch and run `npm run start` from project root and `npm run dev --sync` from `wp-calypso/apps/editing-toolkit` ensuring you're also logged into your sandbox environment.
2. Go to `/new` and go through the process to create a new site, you will land in the editing experience.
3. Sandbox your newly created site in your host file
4. Flush your cache & DNS and then refresh your site made in step 1 and ensure that you're using the `dev` version of the Editing Toolkit
5. Click "Launch". This should open the launch modal. Proceed to the step "Select a domain" and confirm "Use a domain I own" is not present in the domain picker.

- [ ] "Use a domain I own" component is not present


Fixes https://github.com/Automattic/wp-calypso/issues/47372

tjcafferkey committed:
DomainIOwn component

tjcafferkey committed:
Add useYourDomain prop to DomainPicker and add DomainIOwn component to render if prop is true and has a existingSubdomain

tjcafferkey committed:
Enable useYourDomain prop for focused launch domain details step

a8ci18n commented on packages/domain-picker/src/domain-picker/domain-i-own-item.tsx:
```diff
@@ -0,0 +1,64 @@
+				</span>
+				<div>
+					<span className="domain-picker__item-tip">
+						{ __( "You can use it as your site's address", __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'You can use it as your site's address.', { context: 'Upgrades: Register domain description'} )` *ES Score: 8*

tjcafferkey committed:
Change DomainIOwn containing element to be label so it inherits the border-radius of being last-of-type when in the domains suggestions list

tjcafferkey committed:
Add optional redirectTo prop

tjcafferkey committed:
Get Calypso url in Focused Launched

Needed to pass as the source attribute when sending users to the "Use a domain I own" flow

tjcafferkey committed:
Remove target="_blank" attribute from "Use a domain I own" link

tjcafferkey committed:
Remove unnecessary redirectTo param

tjcafferkey committed:
Remove unnecessary redirectTo param

tjcafferkey commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -383,6 +383,7 @@ class CalypsoifyIframe extends Component<
 			const isExperimental = config.isEnabled( 'gutenboarding/feature-picker' ); // TODO: remove after ETK 2.8.6 is released
 			const isPersistentLaunchButton = config.isEnabled( 'create/persistent-launch-button' ); // TODO: remove after ETK 2.8.6 is released
 			const isFocusedLaunchFlow = config.isEnabled( 'create/focused-launch-flow' );
+			const currentCalypsoUrl = window.location.href;
```
@razvanpapadopol is piggy backing on this to get the `currentCalypsoUrl` in the Focused Launch context OK or should we separate this out into a more explicit bit of code do you think?

razvanpapadopol replied:
It seems like we're already transmitting this kind of data [a few lines below](https://github.com/Automattic/wp-calypso/blob/trunk/client/gutenberg/editor/calypsoify-iframe.tsx#L425-L430), information that is then used on the other side in [getCalypsoUrlInfo function](https://github.com/Automattic/wp-calypso/blob/trunk/apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js#L855).

It's shouldn't be a blocker for this PR if it works well. Just make sure it's documented well, at least in code, that we cut some corners.

tjcafferkey committed:
Domain I own component tests

tjcafferkey committed:
Update package name to be valid

razvanpapadopol commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/utils.ts:
```diff
@@ -24,6 +24,14 @@ export const redirectToWpcomPath = ( url: string ) => {
 	redirectParentWindow( `${ origin }${ path }` );
 };
 
+export const getCurrentLaunchFlowUrl = (): string | undefined => {
+	try {
+		return window?.calypsoifyGutenberg?.currentCalypsoUrl as string;
+	} catch ( err ) {
+		return undefined;
+	}
+};
+
```
Isn't the Calypso data already available by the time we execute this code inside the iframe? At least this is what I'd assume by looking at `apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts`.

```suggestion
export const currentLaunchFlowUrl = window?.calypsoifyGutenberg?.currentCalypsoUrl;
```

There is no need to typecast if we add `currentCalypsoUrl?: string;` at line 13 above.

Also I think it's fine to use `window.location.href` as a fallback here in the context of #47803

tjcafferkey replied:
>Isn't the Calypso data already available by the time we execute this code inside the iframe?

The suggested changes don't return the Calypso URL so I wouldn't assume so, but I will revisit this to investigate further.

razvanpapadopol commented on packages/domain-picker/src/domain-picker/style.scss:
```diff
@@ -449,3 +454,17 @@ $accent-blue: #117ac9;
 	}
 }
+
+.domain-picker__action-link {
+	display: flex;
+	align-items: center;
+
+	&.with-underline {
+		text-decoration: underline;
+	}
+
+	a,
+	a:hover {
+		color: var(--mainColor);
+	}
+}
```
I'm looking at this `__action_link` and at the `GoBackButton` in the `@automattic/launch` package and they seem to be styled the same in Figma. Maybe it makes sense for them to share the same CSS and markup (with a prop `arrow: 'left' | 'right'`) and be exported from the `@automattic/onboarding` package?

It's also fine to leave that extraction for later (add a comment in code linking to this comment) and just [copy the CSS](https://github.com/Automattic/wp-calypso/blob/trunk/packages/launch/src/focused-launch/go-back-button/style.scss) to set the correct hover opacity, background, border, and other resets.

tjcafferkey replied:
I've moved this to a component and created an issue to replace the Back Buttons with the newly created component https://github.com/Automattic/wp-calypso/issues/48568

razvanpapadopol commented on packages/domain-picker/src/domain-picker/domain-i-own-item.tsx:
```diff
@@ -0,0 +1,63 @@
+					</span>
+				</div>
+			</div>
+			<div className="domain-picker__action-link with-underline">
```
`with-underline` class doesn't include the namespace. See https://github.com/Automattic/wp-calypso/pull/47062#discussion_r517293404
```suggestion
			<div className="domain-picker__action-link domain-picker__action-link--with-underline">
```

tjcafferkey replied:
Yep this would be preferred but didn't realise we had agreed on a convention.

razvanpapadopol commented on packages/domain-picker/src/domain-picker/domain-i-own-item.tsx:
```diff
@@ -0,0 +1,63 @@
+				</div>
+			</div>
+			<div className="domain-picker__action-link with-underline">
+				{ createInterpolateElement(
+					__(
+						'<strong><use_domain_link>Use a domain I own</use_domain_link></strong>',
+						__i18n_text_domain__
+					),
+					{
+						strong: <strong />,
+						use_domain_link: (
+							<a
+								rel="noreferrer"
+								href={ localizeUrl(
+									`https://wordpress.com/start/new-launch/domains-launch/use-your-domain?siteSlug=${ siteSlug }&source=${ source }`
+								) }
+							/>
+						),
+					}
+				) }
```
Is there a reason for using `createInterpolateElement` here?

```suggestion
				<strong>
					<a
						rel="noreferrer"
						href={ localizeUrl(
							`https://wordpress.com/start/new-launch/domains-launch/use-your-domain?siteSlug=${ siteSlug }&source=${ source }`
						) }
					>
						{ __( 'Use a domain I own', __i18n_text_domain__ ) }
					</a>
				</strong>
```

razvanpapadopol commented on packages/domain-picker/src/domain-picker/domain-i-own-item.tsx:
```diff
@@ -0,0 +1,63 @@
+	source: string;
+}
+
+const DomainIOwn: React.FunctionComponent< Props > = ( { siteSlug, source } ) => {
```
I know there is no linting rule for this but can the Component name [match the filename](https://airbnb.io/javascript/react/#naming)? Of course, the filename should remain in hyphen-case as per our convention.
Also, speaking of naming, `UseYourDomain` wouldn't be more relevant since it matches the name of the flow this component is linking to, at least for now?

razvanpapadopol commented on packages/domain-picker/src/domain-picker/domain-i-own-item.tsx:
```diff
@@ -0,0 +1,63 @@
+		<label className="domain-picker__suggestion-item contains-link">
+			<div className="domain-picker__suggestion-item-name">
+				<span className="domain-picker__domain-wrapper with-margin">
+					{ createInterpolateElement(
+						__( '<strong>Already own a domain?</strong>', __i18n_text_domain__ ),
+						{
+							strong: <strong />,
+						}
+					) }
```
same as https://github.com/Automattic/wp-calypso/pull/48119/files#r547204888

And as a side note, do we really want to use `<strong />` at all? Recently we've been using `font-weight: 500` as an exception for these bolded texts. Eg: https://github.com/Automattic/wp-calypso/blob/trunk/packages/domain-picker/src/domain-picker/style.scss#L270

tjcafferkey committed:
currentCalypsoUrl comments

tjcafferkey committed:
Add getCurrentLaunchFlowUrl type to calypsoifyGutenberg and return window.location.href within catch

tjcafferkey committed:
Update action-link to include namespaced modifier

tjcafferkey committed:
Remove createInterpolateElement where its not needed for the domain I own item

tjcafferkey committed:
Update DomainIOwn naming to be UseYourDomainItem instead

tjcafferkey committed:
Replace createInterpolateElement and use modifier to bold text

a8ci18n commented on packages/domain-picker/src/domain-picker/use-your-domain-item.tsx:
```diff
@@ -0,0 +1,53 @@
+					</span>
+					<div>
+						<span className="domain-picker__item-tip">
+							{ __( "You can use it as your site's address", __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'You can use it as your site's address.', { context: 'Upgrades: Register domain description'} )` *ES Score: 8*

a8ci18n commented on packages/domain-picker/src/domain-picker/use-your-domain-item.tsx:
```diff
@@ -0,0 +1,53 @@
+								`https://wordpress.com/start/new-launch/domains-launch/use-your-domain?siteSlug=${ siteSlug }&source=${ source }`
+							) }
+						>
+							{ __( 'Use a domain I own', __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'Use a domain I own', { context: 'Domain transfer or mapping suggestion button'} )` *ES Score: 8*

tjcafferkey committed:
Add ButtonArrow to onboarding package and replace domain I own link with new component

tjcafferkey committed:
add onboarding to domain picker tsconfig references

a8ci18n commented on packages/domain-picker/src/domain-picker/use-your-domain-item.tsx:
```diff
@@ -0,0 +1,36 @@
+				</span>
+				<div>
+					<span className="domain-picker__item-tip">
+						{ __( "You can use it as your site's address", __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'You can use it as your site's address.', { context: 'Upgrades: Register domain description'} )` *ES Score: 8*

a8ci18n commented on packages/domain-picker/src/domain-picker/use-your-domain-item.tsx:
```diff
@@ -0,0 +1,36 @@
+				</div>
+			</div>
+			<ArrowButton arrow="right" onClick={ onClick }>
+				{ __( 'Use a domain I own', __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'Use a domain I own', { context: 'Domain transfer or mapping suggestion button'} )` *ES Score: 8*

tjcafferkey committed:
Fix component tests for <UseYourDomainItem />

tjcafferkey committed:
Update import path for domain I own

razvanpapadopol commented on packages/domain-picker/src/domain-picker/index.tsx:
```diff
@@ -376,6 +381,9 @@ const DomainPicker: FunctionComponent< Props > = ( {
 											times( placeholdersCount, ( i ) => (
 												<SuggestionItemPlaceholder type={ itemType } key={ i } />
 											) ) }
+										{ ! areDependenciesLoading && existingSubdomain && onUseYourDomainClick && (
```
I'm not sure about the `areDependenciesLoading` check but we definitely don't want `existingSubdomain` to be a condition for displaying this button.
 
```suggestion
										{ ! areDependenciesLoading && onUseYourDomainClick && (
```

tjcafferkey replied:
@razvanpapadopol you are right. I have made some markup and conditional changes around this code but just to explain why:

`areDependenciesLoading` - When domain suggestions are loading, we don't need to perform this check for "Use a domain I own" because they are not a dependency on this item similar to existing sub-domains.
`existingSubDomain` - Not sure how this snuck in, or how I left it but you're right and it's been removed.

I've also amended the CSS selectors for `border-radius` on the first and last items in [this commit](https://github.com/Automattic/wp-calypso/pull/48119/commits/7dd94e90d227f63e4f187758d9f7501d0253515f) and also wrapped the listed domain items in a container because of the following reasons:

* Placeholders and domain suggestions have different "types" (div/label) so the last placeholder isn't actually "last" because "Use a domain I own" is below it. Changing the selector to `last-child` fixed this.
* The "Show all domains" was in the same list so was also being considered as the `last-child` meaning I had to add a containing element to the domain suggestion list to reliably target the actual first and last items.
* I could have just moved "Show all domains" outside of its parent element but it would have created a flexbox bug in the scenario where `showDomainCategories` was `true` because the `flex-direciton` must be `row` to accommodate domain categories.

razvanpapadopol replied:
Thanks, @tjcafferkey, for the fixes and for providing the relevant context 👍 

> last placeholder isn't actually "last" because "Use a domain I own" is below it. Changing the selector to last-child fixed this.

good catch!

Just one more note: after rebase, please make sure the fixes from https://github.com/Automattic/wp-calypso/pull/48172 merged yesterday are still in place.

tjcafferkey replied:
@razvanpapadopol  Done 👍🏻

tjcafferkey committed:
Amend CSS selector types because placeholders / suggested items / domain I own items have different element types.

tjcafferkey committed:
Remove commented out flexbox direction property

razvanpapadopol reviewed (approved):
LGTM! 🙌 
When merging this we need to make sure we deploy also `wpcom-block-editor` immediately so the first step would be to create a diff for that.

tjcafferkey commented:
WPCOM Block Editor patch D55196-code

cc: @razvanpapadopol
